jobbergate-core/jobbergate_core/auth/core.py

```diff
@@ -0,0 +1,316 @@
+        ...     login_audience="https://local.omnivector.solutions",
+        ...     login_client_id="cli",
+        ... )
+        >>> jobbergate_base_url = "http://localhost:8000/jobbergate"
+        >>> jobbergate_auth.acquire_tokens()
+        Login Here: http://keycloak.local:8080/realms/jobbergate-local/device?user_code=LMVJ-XOLG
+        >>> response = requests.get(
+        ...     f"{jobbergate_base_url}/applications",
+        ...     auth=jobbergate_auth # this is the important part
+        )
+        >>> response.raise_for_status()
+        >>> print(f"response = {response.json()}")
+    """
+
+    cache_directory: Path
+    login_domain: str
+    login_audience: str
+    login_client_id: str
+    tokens: Dict[TokenType, Token] = field(default_factory=dict)
+
+    def __call__(self, request):
+        """
+        This internal method allows the integration with the ``requests`` library.
+
+        It is called automatically when the instance is passed to the ``auth`` parameter,
+        see the examples in the class docstring.
+
+        It adds the ``Authorization`` header to the request with the access token.
+        """
+        logger.debug("Authenticating request")
+
+        self.acquire_tokens()
+        access_token = self.check_credentials(token_type=TokenType.ACCESS)
+
+        request.headers["Authorization"] = f"Bearer {access_token.content}"
+        return request
+
+    def acquire_tokens(self):
+        """
+        High-level method to acquire a valid access token.
+
+        This method will attempt, in order:
+
+        * Load the tokens from the cache directory (see :meth:`JobbergateAuth.load_from_cache`)
+        * If the access token is unavailable or expired, refresh both tokens
+          using the refresh token (see :meth:`JobbergateAuth.refresh_tokens`)
+        * If the refresh token is unavailable or expired, login to fetch both tokens
+          (see :meth:`JobbergateAuth.login`)
+        """
+        logger.debug("Acquiring tokens")
+
+        self.load_from_cache(skip_loaded=True)
+        if TokenType.ACCESS in self.tokens and not self.tokens[TokenType.ACCESS].is_expired():
+            return
+        elif TokenType.REFRESH in self.tokens and not self.tokens[TokenType.REFRESH].is_expired():
+            self.refresh_tokens()
+            return
+        self.login()
+
+    def check_credentials(self, token_type: TokenType = TokenType.ACCESS) -> Token:
+        """
+        Check if the credentials on the provided ``token_type`` are available and not expired.
+
+        Arguments:
+            token_type (TokenType): The type of token to check (default is ``access``).
+
+        Returns:
+            Token: The token is returned for reference.
+
+        Raises:
+            AuthenticationError: If the credentials are invalid.
+        """
+        token = cast(Token, self.tokens.get(token_type))
+        AuthenticationError.require_condition(token, f"{token_type} token was not found")
```

Comment:
technically saying, I see some mismatch between the docstring and the line 139. You're checking just if the `token` variable is equivalent to the literal `True`, however, the docstring implies that the function will check the token availability.

Token availability, as far as I understand, means the access token is valid. Do you think the same? If so, then the function should check if the token has the proper content that makes it available as an access token.

For example, the token `whatever` would pass the assertion but it isn't available for usage.

A proper verification here, IMO, is to use armasec to verify if the token can be authenticated against the input domain.

Thoughts?

Reply:
By availability, I mean the given `token_type` is on the internal dictionary that stores them. Since the dictionary starts empty, without `load_from_cache` and/or `login` they are not available.

The current implementation relies on `dict.get()` returning `None` for a key that is not there, it is a bit confusing indeed. I suggest the following change to clarify this:

```
        AuthenticationError.require_condition(token_type in token, f"{token_type} token was not found")
        token = self.tokens[token_type]
```

Regarding the check for tokens' expiration, some applications relying on them may run for a long time, I mean more than their lifespan, so I think it is good to check it at runtime when they are needed. In this way, automatic refresh can take place, for instance.

The point about checking if they are valid using armasec is also good. I see situations like the users changing environments (prod/staging/QA) when the tokens will be available and will be inside their lifespan, but won't be able to access the resources. How difficult do you think it would be? Could it be addressed in a new task instead of in this PR?

Reply:
most likely in other PR I think, specially because the `TokenSecurity` class from Armasec depends on the FastAPI request, which is not the case here. Possibly we would need to copy some code from there instead of pluging Armasec directly

Reply:
Code readability improved in [7f05745](https://github.com/omnivector-solutions/jobbergate/pull/228/commits/7f05745d84b9f1e70e95ecdb1d3bb1dafc0586eb)

Reply:
More details were added on [e7ebf7b](https://github.com/omnivector-solutions/jobbergate/pull/228/commits/e7ebf7b7e4ce8687c48b89c759ea255e98e7e58c)